src/ir/ir_traversing_visitor.h

```diff
@@ -24,47 +25,68 @@ namespace raksha::ir {
 // A visitor that also traverses the children of a node and allows performing
 // different actions before (PreVisit) and after (PostVisit) the children are
 // visited. Override any of the `PreVisit` and `PostVisit` methods as needed.
-template <typename Derived>
-class IRTraversingVisitor : public IRVisitor<Derived> {
+template <typename Derived, typename Result=std::monostate>
+class IRTraversingVisitor : public IRVisitor<Derived, Result> {
```

Comment:
Is it useful to give some way of specifying a default value for the returned result that you can pass before visiting starts? For example, if the point of the visitor is to compute a set, you might want to make the default result the empty set or a set containing some specific element. You might want to make this default result value something that subclasses of `IRTraversingVisitor` can override or define some other way.

Or do you have this somehow and I missed it?

Reply:
This is what I hope to get from defaulting to returning the default constructed value. This will give us empty set/vector/maps/zeros etc. for many different types without any override.

Reply:
Noting here as well just for completeness:

This is also supported by overriding the method `PreVisit`. i.e.

```
template<typename T>
Return PreVisit(const T& _ignored) {
    return Return(/*special arguments*/);
}
```

The benefit of this version, is that the return can be based on `_ignored` and can be different for different structures (e.g. `Module`, `Block`, `Operation`).

I hope I've covered everything you've asked about.

I think this PR is now ready for a full review.